docs: added imageSizes documentation (#160)
front-commerce/developers.front-commerce.com: source/docs/advanced/production-ready/media-middleware.md

It will also lazyload the image. However, in some cases you might not want this to happen. For instance, you don't want this to happen on the main image of your product page. To disable lazyloading, you can use the option `dangerouslyDisableLazyLoad` or (even better in this example) **the `priority` prop that will also add a `<link rel="preload">` to the page.**

> For a more data-efficient browser preload, we also recommend that you define the [`sizes` HTML attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-sizes) for the image. A more detailed explanation about the `sizes` attribute and how to determine the best `sizes` for your image check out [this](https://ericportis.com/posts/2014/srcset-sizes/#part-2) article. Also see [below](#image-sizes) how we added `sizes` in our themes and how and when you should override it. Here is an simple example of the `sizes` attribute:

Reviewer: What we've tried to achieve across the documentation was:
- keep sentences short
- avoid as much as possible adjectives such as "easy", "simple"… and explain things like if the person didn't know the basics (often linking to an existing page is enough)
- use meaningful text for links (instead of "this", "here"…)

I tried to suggest some minor changes to illustrate this.

```suggestion
For a more data-efficient browser preload, we also recommend that you define the [`sizes` HTML attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-sizes) for the image. A more detailed explanation about the `sizes` attribute and how to determine the best `sizes` for your image check out this [srcset + sizes = AWESOME!](https://ericportis.com/posts/2014/srcset-sizes/#part-2) article. [The "Images size" section below](#image-sizes) explains how we added `sizes` attributes in our themes, and how and when you should override it. Here is an example of the `sizes` attribute:
```

/>
```

## Image Sizes

> First be sure you have a good understanding of how the browser selects what image to load depending on the image rendered width and the image widths available to it in the srcset attribute ([this](https://ericportis.com/posts/2014/srcset-sizes/#part-2) link is a good start).

Reviewer: We tend to use `<blockquote>` for these kind of hints. Here is how it could look:

```suggestion
<blockquote class="tip">
**Prerequisites**: be sure you have a good understanding of how the browser selects what image to load. It is a process that depends on the rendered image width and the image widths available in its `srcset` attribute (the [srcset + sizes = AWESOME!](https://ericportis.com/posts/2014/srcset-sizes/#part-2) article is a good explanation).
</blockquote>
```

> We have added sensible defaults to image sizes on key components in the Home, Category, Search, Product pages. Figuring the correct `sizes` to set can be a daunting task, you need to know the different sizes available for your image and take into account the size it will be rendered as in relation to media breakpoints on the page and parent containers' `max-width`, `padding`and`margin` properties. To simplify this process we devised a smart way to determine these numbers in a very straightforward way.

Reviewer: Nit: I think it is better to stay vague here so that we don't have to update it over time ;)
```suggestion
We have added sensible defaults to image sizes on key components in the principal pages. Figuring the correct `sizes` to set can be a daunting task, you need to know the different sizes available for your image and take into account the size it will be rendered as in relation to media breakpoints on the page and parent containers' `max-width`, `padding`and`margin` properties. To simplify this process we devised a smart way to determine these numbers in a very straightforward way.
```

> We have added sensible defaults to image sizes on key components in the Home, Category, Search, Product pages. Figuring the correct `sizes` to set can be a daunting task, you need to know the different sizes available for your image and take into account the size it will be rendered as in relation to media breakpoints on the page and parent containers' `max-width`, `padding`and`margin` properties. To simplify this process we devised a smart way to determine these numbers in a very straightforward way.

Reviewer: This block was dense. Here is a proposal to make it a bit easier to read. Also I think that having a title for the "method" would be good to directly link to it.

```suggestion
We have added sensible defaults to image sizes on key components in the Home, Category, Search, Product pages. Figuring the correct `sizes` to set can be a daunting task. You need to know the different sizes available for your image. You must also take into account the size it will be rendered as, in relation to media breakpoints on the page and parent containers' `max-width`, `padding` and `margin` properties.

### A method to determine image sizes

To simplify this process we devised a method to help you determine these numbers for your project.
```

        "your-image-alt".trim().toLowerCase() &&
      (img.src || "").trim().toLowerCase().indexOf("your-image-src".trim().toLowerCase()) >= 0 &&
      (img?.attributes?.some_custom_prop?.value || "").trim().toLowerCase() ===
        "your-custom-image-prop-value".trim().toLowerCase() &&
      (img.className || "").toLowerCase().indexOf("your-class-name") >= 0
    );
  }
) => {
  const getImage = () => {
    return Array.prototype.filter.call(
      document.getElementsByTagName("img"),
      isMyImage
    )[0];
  };

  const stats = [];
  window.addEventListener("resize", () => {
    const windowSize = window.innerWidth;
    const img = getImage();
    if (!stats.find(([winSize]) => winSize === windowSize)) {
      stats.push([windowSize, img.offsetWidth]);
    }
  });
  return {
    getSizesStats: () => {
      stats.sort(([winSize1], [winSize2]) => winSize1 - winSize2);
      return stats.map((itm) => itm.join("\t")).join("\n");
    },
    getImage,
    clearSizesStats: () => {
      stats.splice(0, stats.length);
    },
    resetIsMyImage: (newIsMyImage) => {
      isMyImage = newIsMyImage;
    },
  };
})();
```

### Image Sizes Example:

> Let's say the data you collected in the [image sizes section above](#image-sizes) are as follows:

Reviewer: This anchor doesn't work. It should be fixed if you accept the above suggestion about a new heading.

1. For the viewport width of 1320 the size of the image becomes larger than 272 (the 272 sized image is not enough in this case). This means for viewport widths above 1320 the 544 image size is needed.
2. For viewport width between 1120 and 1320 the image size is always between 136 and 272. This means for viewport widths above between 1120 and 1320 the 272 image size is sufficient.
3. For viewport width between 1020 and 1120 the image size is larger than 272 again. This means for viewport widths between 1020 and 1120 the 544 image size is needed.
4. For viewport width less than 1020 the image size on the image is always between 136 and 272 again. This means for viewport widths less than 1020 the 272 image size is sufficient.
5. All this translates to the below sizes attribute (p.s. we gave it a 10px buffer):

```jsx
<Image
  ...otherImageProps
  sizes={`
    (min-width: 1310px) 544px
    (min-width: 1120px) 272px
    (min-width: 1010px) 544px
    272px
  `}
/>
```

Now go to the file `CategoryConstants.js` under `./theme-chocolatine/web/theme/pages/Category/` folder. You will find the exact same `sizes` we deduced above there. No magic numbers!

### Image Sizes Defaults

> We have used the method explained above to set default `sizes` to the Home, Category, Search and Product pages. Those defaults found in the Constants file of the respective page are related to the image presets in the `<theme>/config/images.js` and the default values of some scss variables like `$boxSizeMargin`, `$smallContainerWidth` and `$containerWidth` in the `<theme>/web/theme/main.scss` file. So if you have customized any of the default configurations that affect how the image sizes change with viewport width, you should definitely consider adapting the `sizes` values in the Constants files.

Reviewer: ```suggestion
We have used the method explained above to set default `sizes` across the theme. Those defaults found in the Constants file of the respective page are related to the image presets in the `<theme>/config/images.js` and the default values of some SCSS variables like `$boxSizeMargin`, `$smallContainerWidth` and `$containerWidth` in the `<theme>/web/theme/main.scss` file. So if you have customized any of the default configurations that affect how the image sizes change with viewport width, **you should definitely consider adapting the `sizes` values in the Constants files.**
```
